cpp/plugins/cucim.kit.cuslide/src/cuslide/cuslide.cpp

```diff
@@ -147,9 +147,22 @@ static bool CUCIM_ABI parser_parse(CuCIMFileHandle_ptr handle_ptr, cucim::io::fo
     DLDataType dtype{ kDLUInt, 8, 1 };
 
     // TODO: Fill correct values for cucim::io::format::ImageMetadataDesc
-    // TODO: Do not assume channel names as 'RGB'
-    std::pmr::vector<std::string_view> channel_names(
-        { std::string_view{ "R" }, std::string_view{ "G" }, std::string_view{ "B" } }, &resource);
+    uint8_t n_ch = level0_ifd->samples_per_pixel();
+    std::pmr::vector<std::string_view> channel_names(&resource);
+    channel_names.reserve(n_ch);
+    if (n_ch == 3)
+    {
+        channel_names.emplace_back(std::string_view{ "R" });
+        channel_names.emplace_back(std::string_view{ "G" });
+        channel_names.emplace_back(std::string_view{ "B" });
+    }
+    else
```

Comment:
This `else` case also gets run for n_ch = 1 grayscale images, so as-before a single channel image will be listed as having one channel named "R" when printing the metadata from Python

We would still run into problems for any case with > 4 channels (or does TIFF not support that?)

How about using `else if (n_ch == 4)` here and then having a final else clause that just stores a dummy name (maybe "X"?) to all `n_ch` channels?

Reply:
Thanks @grlee77 for the comment!

Didn't check 1-ch image case before. Since loaded image will have 4-ch images when it goes through a slow-path using libtiff, I also updated the output shape of the image.

Reply:
Thanks